File: 1-Divide and Conquer, Sorting and Searching, and Randomized Algorithms/4-Quick Sort/quick_sort.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Quick Sort implementation.
"""

__author__ = 'Ziang Lu'

import random
from typing import List


def quick_sort(nums: List[int]) -> None:
    """
    Sorts the given list using Quick Sort.
    :param nums: list[int]
    :return: None
    """
    # Check whether the input list is null or empty
    if not nums:
        return

    _quick_sort_helper(nums, left=0, right=len(nums) - 1)


def _quick_sort_helper(nums: List[int], left: int, right: int) -> None:
    """
    Private helper function to sort the given sub-list recursively using Quick
    Sort.
    :param nums: list[int]
    :param left: int
    :param right: int
    :return: None
    """
    # Base case
    if left >= right:
        return
    # Recursive case
    # Choose a pivot from the given sub-list, and move it to the left.
    _choose_pivot(nums, left, right)
    pivot_idx = _partition(nums, left=left, right=right)
    _quick_sort_helper(nums, left=left, right=pivot_idx - 1)
    _quick_sort_helper(nums, left=pivot_idx + 1, right=right)


def _choose_pivot(nums: List[int], left: int, right: int) -> None:
    """
    Helper function to choose a pivot from the given sub-list, and move it to
    the left.
    :param nums: list[int]
    :param left: int
    :param right: int
    :return: None
    """
    # [Randomized] Randomly choose a pivot from the given sub-list
    pivot_idx = random.randint(left, right)
    # Move the pivot to the left
    if pivot_idx != left:
        nums[left], nums[pivot_idx] = nums[pivot_idx], nums[left]


def _partition(nums: List[int], left: int, right: int) -> int:
    """
    Helper function to partition the given sub-list.
    :param nums: list[int]
    :param left: int
    :param right: int
    :return: int
    """
    # The pivot has already been moved to the left.
    pivot = nums[left]

    # Iterate over the sub-list, use a pointer to keep track of the smaller
    # part, and swap the current number with the pointer as necessary
    smaller_ptr = left + 1
    i = left + 1
    while True:
        while i <= right and nums[i] > pivot:
            i += 1
        if i > right:
            break
        if i != smaller_ptr:
            nums[smaller_ptr], nums[i] = nums[i], nums[smaller_ptr]
        smaller_ptr += 1
        i += 1
    if left != smaller_ptr - 1:
        nums[left], nums[smaller_ptr - 1] = nums[smaller_ptr - 1], nums[left]
    return smaller_ptr - 1

File: 1-Divide and Conquer, Sorting and Searching, and Randomized Algorithms/4-Quick Sort/QuickSort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

/**
 * Quick Sort implementation.
 *
 * @author Ziang Lu
 */
public class QuickSort {

    /**
     * Sorts the given array using Quick Sort.
     * @param nums array to sort
     */
    public void quickSort(int[] nums) {
        // Check whether the input array is null or empty
        if ((nums == null) || (nums.length == 0)) {
            return;
        }

        quickSortHelper(nums, 0, nums.length - 1);
    }

    /**
     * Private helper method to sort the given sub-array recursively using
     * QuickSort.
     * @param nums array to sort
     * @param left left bound
     * @param right right bound
     */
    private void quickSortHelper(int[] nums, int left, int right) {
        // Base case
        if (left >= right) {
            return;
        }
        // Recursive case
        // Choose a pivot from the given sub-array, and move it to the left
        choosePivot(nums, left, right);
        int pivotIdx = partition(nums, left, right);
        quickSortHelper(nums, left, pivotIdx - 1);
        quickSortHelper(nums, pivotIdx + 1, right);
    }

    /**
     * Helper method to choose a pivot from the given sub-array, and move it to
     * the left.
     * @param nums array to sort
     * @param left left bound
     * @param right right bound
     */
    private void choosePivot(int[] nums, int left, int right, boolean randomly) {
        // [Randomized] Randomly choose a pivot from the given sub-array
        Random random = new Random();
        int pivotIdx = left + random.nextInt(right + 1 - left);
        // [Deterministic] Use the median of medians as the pivot
        // Move the pivot to the left
        if (pivotIdx != left) {
            swap(nums, pivotIdx, left);
        }
    }

    /**
     * Helper method to swap two elements in the given array.
     * @param array given array
     * @param i index of the first element
     * @param j index of the second element
     */
    private void swap(int[] array, int i, int j) {
        int tmp = array[j];
        array[j] = array[i];
        array[i] = tmp;
    }

    /**
     * Helper method to partition the given part of the array.
     * @param nums array to partition
     * @param left left bound
     * @param right right bound
     * @return
     */
    private int partition(int[] nums, int left, int right) {
        // The pivot has already been moved to the left.
        int pivot = nums[left];

        // Iterate over the sub-array, use a pointer to keep track of the smaller part, and swap the current number with
        // the pointer as necessary
        int smallerPtr = left + 1;
        int i = left + 1;
        while (true) {
            while ((i <= right) && (nums[i] > pivot)) {
                ++i;
            }
            if (i > right) {
                break;
            }
            if (i != smallerPtr) {
                swap(nums, i, smallerPtr);
            }
            ++smallerPtr;
            ++i;
        }
        if (left != (smallerPtr - 1)) {
            swap(nums, left, smallerPtr - 1);
        }
        return smallerPtr - 1;
    }

}
